In [149]:
print('HIIIiii')

HIIIiii


In [150]:
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
%matplotlib inline
import matplotlib
matplotlib.rcParams["figure.figsize"] = (20,10)

In [151]:
df1 = pd.read_csv("bengaluru_house_prices.csv")
df1.head()

area_type   availability                  location       size  \
0  Super built-up  Area         19-Dec  Electronic City Phase II      2 BHK   
1            Plot  Area  Ready To Move          Chikka Tirupathi  4 Bedroom   
2        Built-up  Area  Ready To Move               Uttarahalli      3 BHK   
3  Super built-up  Area  Ready To Move        Lingadheeranahalli      3 BHK   
4  Super built-up  Area  Ready To Move                  Kothanur      2 BHK   

   society total_sqft  bath  balcony   price  
0  Coomee        1056   2.0      1.0   39.07  
1  Theanmp       2600   5.0      3.0  120.00  
2      NaN       1440   2.0      3.0   62.00  
3  Soiewre       1521   3.0      1.0   95.00  
4      NaN       1200   2.0      1.0   51.00

In [152]:
df1.shape

(13320, 9)

In [153]:
df1.groupby('area_type')['area_type'].agg('count')

area_type
Built-up  Area          2418
Carpet  Area              87
Plot  Area              2025
Super built-up  Area    8790
Name: area_type, dtype: int64

In [154]:
df2 = df1.drop(['bath','balcony','availability','society'],axis='columns') #May not be advisible to drop these parameters in the real life scenarios
df2.tail()

area_type               location       size total_sqft  \
13315        Built-up  Area             Whitefield  5 Bedroom       3453   
13316  Super built-up  Area          Richards Town      4 BHK       3600   
13317        Built-up  Area  Raja Rajeshwari Nagar      2 BHK       1141   
13318  Super built-up  Area        Padmanabhanagar      4 BHK       4689   
13319  Super built-up  Area           Doddathoguru      1 BHK        550   

       price  
13315  231.0  
13316  400.0  
13317   60.0  
13318  488.0  
13319   17.0

In [155]:
df2.shape


(13320, 5)

# Data Cleaning

Identifying and Removing the null or empty entries


In [156]:
df2.isnull().sum() 

area_type      0
location       1
size          16
total_sqft     0
price          0
dtype: int64

In [157]:
df3 = df2.dropna()
df3.isnull().sum()

area_type     0
location      0
size          0
total_sqft    0
price         0
dtype: int64

Unifying the Size attribute

In [185]:
df3['size'].unique()

array(['2 BHK', '4 Bedroom', '3 BHK', '4 BHK', '6 Bedroom', '3 Bedroom',
       '1 BHK', '1 RK', '1 Bedroom', '8 Bedroom', '2 Bedroom',
       '7 Bedroom', '5 BHK', '7 BHK', '6 BHK', '5 Bedroom', '11 BHK',
       '9 BHK', '9 Bedroom', '27 BHK', '10 Bedroom', '11 Bedroom',
       '10 BHK', '19 BHK', '16 BHK', '43 Bedroom', '14 BHK', '8 BHK',
       '12 Bedroom', '13 BHK', '18 Bedroom'], dtype=object)

In [159]:
df3['size'].nunique()

31

In [160]:
roney = lambda a, b : a+b
print(roney(68,1))

69


In [161]:
df3['bhk'] = df3['size'].apply(lambda x: int(x.split(' ')[0]))
df3.head(10)

C:\Users\Roney Patel\AppData\Local\Temp\ipykernel_14200\3668831979.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df3['bhk'] = df3['size'].apply(lambda x: int(x.split(' ')[0]))


area_type                  location       size total_sqft  \
0  Super built-up  Area  Electronic City Phase II      2 BHK       1056   
1            Plot  Area          Chikka Tirupathi  4 Bedroom       2600   
2        Built-up  Area               Uttarahalli      3 BHK       1440   
3  Super built-up  Area        Lingadheeranahalli      3 BHK       1521   
4  Super built-up  Area                  Kothanur      2 BHK       1200   
5  Super built-up  Area                Whitefield      2 BHK       1170   
6  Super built-up  Area          Old Airport Road      4 BHK       2732   
7  Super built-up  Area              Rajaji Nagar      4 BHK       3300   
8  Super built-up  Area              Marathahalli      3 BHK       1310   
9            Plot  Area              Gandhi Bazar  6 Bedroom       1020   

    price  bhk  
0   39.07    2  
1  120.00    4  
2   62.00    3  
3   95.00    3  
4   51.00    2  
5   38.00    2  
6  204.00    4  
7  600.00    4  
8   63.25    3  
9  370.00    6

In [162]:
df3['bhk'].unique(), df3.drop(['size'], axis='columns'), df3.shape,  

(array([ 2,  4,  3,  6,  1,  8,  7,  5, 11,  9, 27, 10, 19, 16, 43, 14, 12,
        13, 18]),
                   area_type                  location total_sqft   price  bhk
 0      Super built-up  Area  Electronic City Phase II       1056   39.07    2
 1                Plot  Area          Chikka Tirupathi       2600  120.00    4
 2            Built-up  Area               Uttarahalli       1440   62.00    3
 3      Super built-up  Area        Lingadheeranahalli       1521   95.00    3
 4      Super built-up  Area                  Kothanur       1200   51.00    2
 ...                     ...                       ...        ...     ...  ...
 13315        Built-up  Area                Whitefield       3453  231.00    5
 13316  Super built-up  Area             Richards Town       3600  400.00    4
 13317        Built-up  Area     Raja Rajeshwari Nagar       1141   60.00    2
 13318  Super built-up  Area           Padmanabhanagar       4689  488.00    4
 13319  Super built-up  Area         

In [163]:
df3[df3.bhk>20], df3.bhk>20

(                 area_type                   location        size total_sqft  \
 1718  Super built-up  Area  2Electronic City Phase II      27 BHK       8000   
 4684            Plot  Area                Munnekollal  43 Bedroom       2400   
 
       price  bhk  
 1718  230.0   27  
 4684  660.0   43  ,
 0        False
 1        False
 2        False
 3        False
 4        False
          ...  
 13315    False
 13316    False
 13317    False
 13318    False
 13319    False
 Name: bhk, Length: 13303, dtype: bool)

In [165]:
def is_float(x):
    try:
        float(x)
    except:
        return False
    return True

In [166]:
df3[~df3.total_sqft.apply(is_float)].head(10)  #Use negation so that the entries with output of False i.e other than integers are displayed

area_type            location       size      total_sqft  \
30   Super built-up  Area           Yelahanka      4 BHK     2100 - 2850   
56         Built-up  Area         Devanahalli  4 Bedroom     3010 - 3410   
81         Built-up  Area         Hennur Road  4 Bedroom     2957 - 3450   
122  Super built-up  Area              Hebbal      4 BHK     3067 - 8156   
137  Super built-up  Area  8th Phase JP Nagar      2 BHK     1042 - 1105   
165  Super built-up  Area            Sarjapur      2 BHK     1145 - 1340   
188  Super built-up  Area            KR Puram      2 BHK     1015 - 1540   
224  Super built-up  Area         Devanahalli      3 BHK     1520 - 1740   
410  Super built-up  Area             Kengeri      1 BHK  34.46Sq. Meter   
549  Super built-up  Area         Hennur Road      2 BHK     1195 - 1440   

       price  bhk  
30   186.000    4  
56   192.000    4  
81   224.500    4  
122  477.000    4  
137   54.005    2  
165   43.490    2  
188   56.800    2  
224   74.820    3  
410   18.500    1  
549   63.770    2

Look out at the ninth entry with the index of 410, it is in sq meter as well as has string input in the cell

In [167]:
def convert_range(x):
    token = x.split('-')
    if len(token) == 2:
        return (float(token[0])+ float(token[1]))/2
    try:
        return float(x)
    except:
        return None

In [168]:
df4 = df3.drop(['size'],axis='columns')
df4.total_sqft = df4.total_sqft.apply(convert_range)
df4[df4.location == 'Hebbal'], print('\n'), df4[df4.index == 410]

(                  area_type location  total_sqft  price  bhk
 32     Super built-up  Area   Hebbal      1760.0  123.0    3
 122    Super built-up  Area   Hebbal      5611.5  477.0    4
 228    Super built-up  Area   Hebbal      1900.0  119.0    3
 293    Super built-up  Area   Hebbal      1349.0   98.2    2
 370    Super built-up  Area   Hebbal      2483.0  212.0    4
 ...                     ...      ...         ...    ...  ...
 13075        Built-up  Area   Hebbal      2250.0  219.0    3
 13092  Super built-up  Area   Hebbal      4000.0  440.0    4
 13105  Super built-up  Area   Hebbal      1349.0   98.0    2
 13208  Super built-up  Area   Hebbal      4000.0  370.0    4
 13289  Super built-up  Area   Hebbal      1349.0   96.8    2
 
 [177 rows x 5 columns],
 None,
                 area_type location  total_sqft  price  bhk
 410  Super built-up  Area  Kengeri         NaN   18.5    1)

Observe that the entry 122 now has a finite float value


In [169]:
df4.loc[410],   df4[df4.total_sqft.isnull()].count()

(area_type     Super built-up  Area
 location                   Kengeri
 total_sqft                     NaN
 price                         18.5
 bhk                              1
 Name: 410, dtype: object,
 area_type     46
 location      46
 total_sqft     0
 price         46
 bhk           46
 dtype: int64)

# Feature Engineering and Dimensionality Reduction


Lets create a column named price_per_sqft
Given price here is in Lakhs


In [170]:
df4['price_per_sqft'] = df4['price']*100000/df4['total_sqft']
df4.head()

area_type                  location  total_sqft   price  bhk  \
0  Super built-up  Area  Electronic City Phase II      1056.0   39.07    2   
1            Plot  Area          Chikka Tirupathi      2600.0  120.00    4   
2        Built-up  Area               Uttarahalli      1440.0   62.00    3   
3  Super built-up  Area        Lingadheeranahalli      1521.0   95.00    3   
4  Super built-up  Area                  Kothanur      1200.0   51.00    2   

   price_per_sqft  
0     3699.810606  
1     4615.384615  
2     4305.555556  
3     6245.890861  
4     4250.000000

In [171]:
# df4.location.nunique()
len(df4.location.unique())

1304

GROUP BY FUNCTION

In [172]:
df4.groupby('location')['price_per_sqft'].mean().sort_values(ascending=False)

location
Grihalakshmi Layout                    625000.000000
Srirampuram                            481004.144004
Suragajakkanahalli                     339363.636364
Weavers Colony                          68801.313629
Raghuvanahalli                          38647.926429
                                           ...      
5 Bedroom Farm House in Lakshmipura              NaN
Chickpet                                         NaN
Marathi Layout                                   NaN
Narasapura                                       NaN
arudi                                            NaN
Name: price_per_sqft, Length: 1304, dtype: float64

In [173]:
df4[df4.price_per_sqft == df4.price_per_sqft.max()]

area_type        location  total_sqft  price  bhk  price_per_sqft
4086  Plot  Area  Sarjapur  Road         1.0  120.0    4      12000000.0

Found an anomaly in DataSet because 1 sqft cant be priced at 1.2Cr.

In [174]:
df4.nlargest(20, 'price_per_sqft')

area_type             location  total_sqft   price  bhk  \
4086             Plot  Area       Sarjapur  Road         1.0   120.0    4   
4972         Built-up  Area          Srirampuram         5.0   115.0    7   
349              Plot  Area   Suragajakkanahalli        11.0    74.0    3   
1122         Built-up  Area  Grihalakshmi Layout        24.0   150.0    5   
11558            Plot  Area           Whitefield        60.0   218.0    4   
1020           Carpet  Area       Weavers Colony        15.0    30.0    1   
7657         Built-up  Area       Raghuvanahalli       425.0   750.0    1   
7088         Built-up  Area          Srirampuram       650.0   500.0    1   
6421             Plot  Area         Bommenahalli      2940.0  2250.0    4   
12443            Plot  Area       Dollars Colony      4350.0  2600.0    4   
8398   Super built-up  Area    Bannerghatta Road      2500.0  1400.0    5   
7883   Super built-up  Area        Vasanth nagar      2000.0  1063.0    4   
9535             Plot  Area         Indira Nagar      2400.0  1250.0    4   
5970             Plot  Area          Mysore Road        45.0    23.0    1   
13067            Plot  Area       Defence Colony      7150.0  3600.0   10   
1567             Plot  Area          Uttarahalli       400.0   200.0    5   
5042             Plot  Area  2nd Block Jayanagar       825.0   400.0    6   
9195   Super built-up  Area           Whitefield      1496.0   718.0    3   
9737         Built-up  Area          Ngef Layout      2100.0  1000.0    4   
13305          Carpet  Area             Hulimavu       500.0   220.0    1   

       price_per_sqft  
4086     1.200000e+07  
4972     2.300000e+06  
349      6.727273e+05  
1122     6.250000e+05  
11558    3.633333e+05  
1020     2.000000e+05  
7657     1.764706e+05  
7088     7.692308e+04  
6421     7.653061e+04  
12443    5.977011e+04  
8398     5.600000e+04  
7883     5.315000e+04  
9535     5.208333e+04  
5970     5.111111e+04  
13067    5.034965e+04  
1567     5.000000e+04  
5042     4.848485e+04  
9195     4.799465e+04  
9737     4.761905e+04  
13305    4.400000e+04

In [175]:
df4.nsmallest(3, 'price_per_sqft'), df4.nlargest(1, 'price_per_sqft'), 

(                 area_type          location  total_sqft  price  bhk  \
 1894            Plot  Area       Nelamangala     52272.0  140.0    3   
 674         Built-up  Area         Yelahanka     35000.0  130.0    3   
 5393  Super built-up  Area  Doddabommasandra     42000.0  175.0    9   
 
       price_per_sqft  
 1894      267.829813  
 674       371.428571  
 5393      416.666667  ,
        area_type        location  total_sqft  price  bhk  price_per_sqft
 4086  Plot  Area  Sarjapur  Road         1.0  120.0    4      12000000.0)

In [176]:
df4.groupby('location')['location'].count().sort_values(ascending=False)

location
Whitefield                                            539
Sarjapur  Road                                        397
Electronic City                                       302
Kanakpura Road                                        273
Thanisandra                                           234
                                                     ... 
bsk 6th stage 2ad block near sri conversation hall      1
beml layout, basaveshwara nagar                         1
basaveshwarnagar                                        1
banashankari stage iii sa                               1
asha township, off hennur road                          1
Name: location, Length: 1304, dtype: int64

In [177]:
df4.location.value_counts(ascending=True)

location
Subhash Nagar                           1
Craig Park Layout                       1
Maruthi HBCS Layout                     1
Brigade Road                            1
Kengeri Satellite Town ( BDA SITE)      1
                                     ... 
Thanisandra                           234
Kanakpura Road                        273
Electronic City                       302
Sarjapur  Road                        397
Whitefield                            539
Name: count, Length: 1304, dtype: int64

In [178]:
loc_stats = df4.groupby('location')['location'].count().sort_values(ascending=False)
len(loc_stats), len(loc_stats[loc_stats<=10])

(1304, 1063)

Out of 1304 entries in the data Set, we have 1063 entries where the location count == 1, hence we club all of these locations that have been appeared only once into a category called 'OTHER'

In [181]:
loc_less_than_10 = loc_stats[loc_stats<=10].index
df4.location = df4.location.apply(lambda x : 'other' if x in loc_less_than_10 else x)

In [183]:
df4.location.value_counts()

location
other                        2897
Whitefield                    539
Sarjapur  Road                397
Electronic City               302
Kanakpura Road                273
                             ... 
Tindlu                         11
Marsur                         11
2nd Phase Judicial Layout      11
Thyagaraja Nagar               11
HAL 2nd Stage                  11
Name: count, Length: 242, dtype: int64

In [187]:
df4.location.nunique()

242